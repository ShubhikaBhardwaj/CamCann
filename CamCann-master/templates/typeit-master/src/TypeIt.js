import Instance from "./Instance";
import allHaveStatus from "./helpers/allHaveStatus";
import { generateHash } from "./utilities";
import toArrayOfNodes from "./helpers/toArrayOfNodes";

export default class TypeIt {
  constructor(element, options) {
    this.instances = toArrayOfNodes(element).map(el => {
      return new Instance({
        typeIt: this,
        element: el,
        id: generateHash(),
        options,
        queue: []
      });
    });
  }

  each(func) {
    this.instances.forEach(instance => {
      func.call(this, instance);
    });
  }

  /**
   * Push a specific action directly into the queue of each instance.
   * If an instance has already completed, trigger the queeu again.
   *
   * @param {string} function
   * @param {*} argument
   * @
   */
  queueUp(action, argument = null, numberOfTimesToCopy = 1) {
    this.each(instance => {
      let isIndependentFunction = typeof action !== "string";

      /**
       * If action is a string, bind it to instance.
       * Otherwise, leave it on its own.
       */
      let toFire = isIndependentFunction ? action : instance[action];

      let toPassAsArguments = isIndependentFunction ? this : argument;

      for (let i = 0; i < numberOfTimesToCopy; i++) {
        instance.queue.add([toFire, toPassAsArguments]);
      }
    });
  }

  is(status) {
    return allHaveStatus(this.instances, status, true);
  }

  freeze() {
    this.each(instance => {
      instance.status.frozen = true;
    });
  }

  unfreeze() {
    this.each(instance => {
      if (!instance.status.frozen) return;
      instance.status.frozen = false;
      instance.fire();
    });
  }

  /**
   * If used after typing has started, will append strings to the end of the existing queue. If used when typing is paused, will restart it.
   *
   * @param  {string} string The string to be typed.
   * @return {object} TypeIt instance
   */
  type(string = "") {
    this.each(instance => instance.queueString(string));
    return this;
  }

  /**
   * If null is passed, will delete whatever's currently in the element.
   *
   * @param  { number } numCharacters Number of characters to delete.
   * @return { TypeIt }
   */
  delete(numberOfCharactersToDelete = null) {
    this.queueUp(
      "delete",
      numberOfCharactersToDelete === null, //-- Maybe delete all.
      numberOfCharactersToDelete === null ? 1 : numberOfCharactersToDelete
    );

    return this;
  }

  pause(ms = null) {
    this.queueUp("pause", ms);
    return this;
  }

  break() {
    this.queueUp("type", document.createElement("BR"));
    return this;
  }

  options(options) {
    this.queueUp("setOptions", options);
    return this;
  }

  exec(func) {
    this.queueUp(func);
    return this;
  }

  /**
   * Destroy the instance, mark as such, and clean up. If specified,
   * remove the cursor on destruction.
   *
   * @param {boolean} removeCursor
   */
  destroy(removeCursor = true) {
    this.each(instance => {
      // Destroy each timeout, for good housekeeping.
      instance.timeouts.forEach(timeout => {
        clearTimeout(timeout);
      });

      instance.timeouts = [];

      // Clean up cursor node, if specified.
      if (removeCursor) {
        let cursorNode = instance.isInput
          ? null
          : instance.$e.querySelector(".ti-cursor");

        if (cursorNode) {
          instance.$e.removeChild(cursorNode);
        }
      }

      instance.status.destroyed = true;
    });
  }

  empty() {
    this.queueUp("empty");
    return this;
  }

  /**
   * Reset each instance like it's brand new.
   */
  reset() {
    this.destroy();

    this.instances = this.instances.map(instance => {
      return instance.reset();
    });

    return this;
  }

  /**
   * Initialize each instance.
   */
  go() {
    this.each(instance => {
      instance.init();
    });

    return this;
  }
}
