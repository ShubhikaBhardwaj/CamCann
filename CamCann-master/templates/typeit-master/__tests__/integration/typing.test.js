import TypeIt from "../../src/TypeIt";

test("Generates a queue correctly.", () => {
  document.body.innerHTML = `<div>'
      <span id="element"></span>
    </div>`;

  const instance = new TypeIt("#element", {
    strings: ["Taxation is...", "theft."]
  }).go();

  instance.instances.forEach(instance => {
    expect(instance.queue).toMatchSnapshot();
  });
});

test("Generates a queue correctly when chaining upon instantiation.", () => {
  document.body.innerHTML = `<div>'
      <span id="element"></span>
    </div>`;

  const instance = new TypeIt("#element", {})
    .type("First string.")
    .delete()
    .type("Second string.")
    .go();

  instance.instances.forEach(instance => {
    expect(instance.queue.waiting).toHaveLength(29);
  });
});

test("Generates correct `nextStringDelay`.", () => {
  document.body.innerHTML = `<div>'
      <span id="element"></span>
    </div>`;

  const instance1 = new TypeIt("#element", {
    nextStringDelay: 500,
    strings: ["Free markets...", "win."]
  }).go();

  let nextStringDelay = instance1.instances[0].opts.nextStringDelay;

  expect(typeof nextStringDelay).toBe("object");
  expect(nextStringDelay.before).toBe(250);
  expect(nextStringDelay.after).toBe(250);
  expect(nextStringDelay.total).toBe(500);

  const instance2 = new TypeIt("#element", {
    nextStringDelay: [150, 400],
    strings: ["Free markets...", "win."]
  }).go();

  nextStringDelay = instance2.instances[0].opts.nextStringDelay;

  expect(nextStringDelay.before).toBe(150);
  expect(nextStringDelay.after).toBe(400);
  expect(nextStringDelay.total).toBe(550);
});

test("Generates correct `loopDelay`.", () => {
  document.body.innerHTML = `<div>'
      <span id="element"></span>
    </div>`;

  const instance1 = new TypeIt("#element", {
    nextStringDelay: 500,
    strings: ["Free markets...", "win."]
  }).go();

  let loopDelay = instance1.instances[0].opts.loopDelay;

  expect(loopDelay).toBe(null);

  const instance2 = new TypeIt("#element", {
    loopDelay: [3000, 5000],
    strings: ["Free markets...", "win."]
  }).go();

  loopDelay = instance2.instances[0].opts.loopDelay;

  expect(typeof loopDelay).toBe("object");
  expect(loopDelay.before).toBe(3000);
  expect(loopDelay.after).toBe(5000);
  expect(loopDelay.total).toBe(8000);
});

test("Removes empty HTML when necessary.", () => {
  document.body.innerHTML = `<div>'
      <span id="element"></span>
    </div>`;

  const instance = new TypeIt("#element", {
    breakLines: false
  })
    .type("This is a string with some <strong>bold.</strong>")
    .delete(5)
    .type("standard text.")
    .delete(14)
    .type("<em>italicized text...</em>")
    .delete(18)
    .type("standard text again.")
    .go();

  let emptyTagPattern = /<[^\/>][^>]*><\/[^>]+>/;

  let result = document.getElementById("element").innerHTML;

  //-- Ensure our regex correctly finds empty tags.
  expect("Text with <strong></strong> a set of empty tags!").toMatch(
    emptyTagPattern
  );

  //-- Will not have any empty HTML.
  expect(result).not.toMatch(emptyTagPattern);
});
