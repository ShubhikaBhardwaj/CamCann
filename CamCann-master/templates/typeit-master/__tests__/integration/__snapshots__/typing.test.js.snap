// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generates a queue correctly. 1`] = `
_default {
  "executed": Array [],
  "waiting": Array [
    Array [
      [Function],
      250,
    ],
    Array [
      [Function],
      "T",
      Object {
        "isFirst": true,
      },
    ],
    Array [
      [Function],
      "a",
    ],
    Array [
      [Function],
      "x",
    ],
    Array [
      [Function],
      "a",
    ],
    Array [
      [Function],
      "t",
    ],
    Array [
      [Function],
      "i",
    ],
    Array [
      [Function],
      "o",
    ],
    Array [
      [Function],
      "n",
    ],
    Array [
      [Function],
      " ",
    ],
    Array [
      [Function],
      "i",
    ],
    Array [
      [Function],
      "s",
    ],
    Array [
      [Function],
      ".",
    ],
    Array [
      [Function],
      ".",
    ],
    Array [
      [Function],
      ".",
      Object {
        "isLast": true,
      },
    ],
    Array [
      [Function],
      375,
    ],
    Array [
      [Function],
      <br />,
    ],
    Array [
      [Function],
      375,
    ],
    Array [
      [Function],
      "t",
      Object {
        "isFirst": true,
      },
    ],
    Array [
      [Function],
      "h",
    ],
    Array [
      [Function],
      "e",
    ],
    Array [
      [Function],
      "f",
    ],
    Array [
      [Function],
      "t",
    ],
    Array [
      [Function],
      ".",
      Object {
        "isLast": true,
      },
    ],
  ],
}
`;
